src/cio_websocket_location_handler.c

```diff
@@ -0,0 +1,297 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) <2017> <Stephan Gatzka>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <ctype.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "cio_base64.h"
+#include "cio_compiler.h"
+#include "cio_http_client.h"
+#include "cio_http_location_handler.h"
+#include "cio_string.h"
+#include "cio_util.h"
+#include "cio_websocket.h"
+#include "cio_websocket_location_handler.h"
+#include "cio_write_buffer.h"
+#include "sha1/sha1.h"
+
+#define CIO_CRLF "\r\n"
+
+enum header_field {
+	CIO_WS_HEADER_UNKNOWN,
+	CIO_WS_HEADER_SEC_WEBSOCKET_KEY,
+	CIO_WS_HEADER_SEC_WEBSOCKET_VERSION,
+	CIO_WS_HEADER_SEC_WEBSOCKET_PROTOCOL,
+};
+
+static enum cio_http_cb_return save_websocket_key(uint8_t *dest, const char *at, size_t length)
+{
+	static const char ws_guid[SEC_WEB_SOCKET_GUID_LENGTH] = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";
+
+	if (length == SEC_WEB_SOCKET_KEY_LENGTH) {
+		memcpy(dest, at, length);
+		memcpy(&dest[length], ws_guid, sizeof(ws_guid));
+		return CIO_HTTP_CB_SUCCESS;
+	} else {
+		return CIO_HTTP_CB_ERROR;
+	}
+}
+
+static enum cio_http_cb_return check_websocket_version(const char *at, size_t length)
+{
+	static const char version[2] = "13";
+	if ((length == sizeof(version)) && (memcmp(at, version, length) == 0)) {
+		return CIO_HTTP_CB_SUCCESS;
+	} else {
+		return CIO_HTTP_CB_ERROR;
+	}
+}
+
+static bool find_requested_sub_protocol(struct cio_websocket_location_handler *handler, const char *name, size_t length)
+{
+	for (unsigned int i = 0; i < handler->number_subprotocols; i++) {
+		const char *sub_protocol = handler->subprotocols[i];
+		size_t name_length = strlen(sub_protocol);
+		if (name_length == length) {
+			if (memcmp(sub_protocol, name, length) == 0) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS1066)